Doc/library/ftplib.rst

```diff
@@ -55,17 +55,43 @@ Reference
 FTP objects
 ^^^^^^^^^^^
 
-.. class:: FTP(host='', user='', passwd='', acct='', timeout=None, source_address=None, *, encoding='utf-8')
-
-   Return a new instance of the :class:`FTP` class.  When *host* is given, the
-   method call ``connect(host)`` is made.  When *user* is given, additionally
-   the method call ``login(user, passwd, acct)`` is made (where *passwd* and
-   *acct* default to the empty string when not given).  The optional *timeout*
-   parameter specifies a timeout in seconds for blocking operations like the
-   connection attempt (if is not specified, the global default timeout setting
-   will be used). *source_address* is a 2-tuple ``(host, port)`` for the socket
-   to bind to as its source address before connecting. The *encoding* parameter
-   specifies the encoding for directories and filenames.
+.. class:: FTP(host='', user='', passwd='', acct='', timeout=None, \
+               source_address=None, *, encoding='utf-8')
+
+   Return a new instance of the :class:`FTP` class.
+   When *host* is given, the method call :meth:`connect(host) <connect>`
+   is made by the constructor.
+   When *user* is given, additionally the method call
+   :meth:`login(user, passwd, acct) <connect>` is made.
+
+   :param str host:
+      The hostname to connect to.
+
+   :param str user:
+      The username to log in with.
+      If empty string, ``"anonymous"`` is used.
+
+   :param str passwd:
+      The password to use when logging in.
+      If not given, and if *passwd* is the empty string or ``"-"``,
+      a password will be automatically generated.
+
+   :param str acct:
+      Account information; see the ACCT FTP command.
+
+   :param timeout:
+      A timeout in seconds for blocking operations like :meth:`connect`.
+      If not specified, the global default timeout setting will be used.
+   :type timeout: int | None
```

Comment:
We just got a bug report at typeshed (in the form of a PR -- the best kind!) arguing that our type hints are incorrect for the `timeout` parameters in `ftplib`, because we allow `float` for this parameter currently whereas the "official docs" say it has to be `int`. Does it _have_ to be a whole number of seconds, or is it okay to pass `timeout=3.14`?

- https://github.com/python/typeshed/pull/11419

Reply:
From the top of my head, I would guess it could be a `float` as well. IIRC, that's what sqlite3.connect()'s *timeout* param also accepts.

Reply:
The FTP class docstring says *timeout* must be numeric.

Reply:
(The typeshed PR also points out that the parameter should accept `None` in our type hints -- skimming through the source code for the module, that _does_ look correct to me. But not sure that `float` should be disallowed.)

Reply:
FTR, *timeout* is passed on to `socket.create_connection`: https://docs.python.org/3/library/socket.html#socket.create_connection

Reply:
... and then to `sock.settimeout`, which accepts a `float`: https://docs.python.org/3/library/socket.html#socket.socket.settimeout

Reply:
Good catch!

Reply:
Amazing, thanks for the quick investigation! I propose that we change the type annotation in the docs for this parameter to `float | None`, and clarify with words that `int` is also accepted as well as `float` (since casual readers of the `ftplib` docs may not be familiar with PEP-484's [numeric tower](https://peps.python.org/pep-0484/#the-numeric-tower))

Reply:
Yeah, it's quite common in the docs elsewhere to say "float" and mean "well, either a float or an int would do". And that's what we'd do in typeshed as well -- type checkers are fine with that kind of thing. So let's do that 👍